File: kotlin/mergeSort.kt
fun merge(leftList: List<Int>, rightRight: List<Int>): List<Int> {
    var indexLeft = 0
    var indexRight = 0
    var newList: MutableList<Int> = mutableListOf()

    while (indexLeft < leftList.count() && indexRight < rightRight.count()) {
        if (leftList[indexLeft] <= rightRight[indexRight]) {
            newList.add(leftList[indexLeft])
            indexLeft++
        } else {
            newList.add(rightRight[indexRight])
            indexRight++
        }
    }

    while (indexLeft < leftList.size) {
        newList.add(leftList[indexLeft])
        indexLeft++
    }

    while (indexRight < rightRight.size) {
        newList.add(rightRight[indexRight])
        indexRight++
    }

    return newList;
}

fun mergeSort(inputList: List<Int>): List<Int> {
    if (inputList.size <= 1) {
        return inputList
    }

    val middle = inputList.size / 2
    var left = inputList.subList(0, middle)
    var right = inputList.subList(middle, inputList.size)

    return merge(mergeSort(left), mergeSort(right))
}


fun main() {
    val input = mutableListOf(5, 4, 3, 2, 1)
    val result = mergeSort(input)
    result.forEach { value ->
        print("$value ")
    }
}

File: PYTHON/Machine Sorting/merge_sort.py
def mergesort(list1):
    if len(list1)>1:
        mid = len(list1)//2
        left_list = list1[:mid]
        right_list = list1[mid:]
        mergesort(left_list)
        mergesort(right_list)
        i = 0
        j = 0
        k = 0
        while i<len(left_list) and j<len(right_list):
            if left_list[i]<right_list[j]:
                list1[k] = left_list[i]
                i=i+1
                k=k+1
            else:
                list1[k] = right_list[j]
                j=j+1
                k=k+1
        while i<len(left_list):
            list1[k] = left_list[i]
            i=i+1
            k=k+1
        while j<len(right_list):
            list1[k] = right_list[j]
            j=j+1
            k=k+1


